include/RAJA/util/SoAPtr.hpp

```diff
@@ -165,6 +168,77 @@ class SoAPtr<RAJA::reduce::detail::ValueLoc<T, IndexType, doing_min>, mempool, a
   second_type* mem_idx = nullptr;
 };
 
+/*!
+ * @brief Specialization for RAJA::expt::ValLoc.
+ */
+template <typename T, typename IndexType, typename mempool, typename accessor>
+class SoAPtr<RAJA::expt::ValLoc<T, IndexType>, mempool, accessor>
+{
+  using first_type = T;
+  using second_type = IndexType;
+
+  template < typename, typename, typename >
+  friend class SoAPtr; // friend other instantiations of this class
+
+public:
+  using value_type = RAJA::expt::ValLoc<T, IndexType>;
+
+  template < typename rhs_accessor >
+  using rebind_accessor = SoAPtr<value_type, mempool, rhs_accessor>;
+
+  SoAPtr() = default;
+  SoAPtr(SoAPtr const&) = default;
+  SoAPtr(SoAPtr &&) = default;
+  SoAPtr& operator=(SoAPtr const&) = default;
+  SoAPtr& operator=(SoAPtr &&) = default;
+
+  explicit SoAPtr(size_t size)
+      : mem(mempool::getInstance().template malloc<first_type>(size)),
+        mem_idx(mempool::getInstance().template malloc<second_type>(size))
+  {
+  }
+
+  template < typename rhs_accessor,
+             std::enable_if_t<!std::is_same<accessor, rhs_accessor>::value>* = nullptr >
+  RAJA_HOST_DEVICE
+  explicit SoAPtr(SoAPtr<value_type, mempool, rhs_accessor> const& rhs)
+    : mem(rhs.mem)
+    , mem_idx(rhs.mem_idx)
+  { }
+
+  SoAPtr& allocate(size_t size)
+  {
+    mem = mempool::getInstance().template malloc<first_type>(size);
+    mem_idx = mempool::getInstance().template malloc<second_type>(size);
+    return *this;
+  }
+
+  SoAPtr& deallocate()
+  {
+    mempool::getInstance().free(mem);
+    mem = nullptr;
+    mempool::getInstance().free(mem_idx);
+    mem_idx = nullptr;
+    return *this;
+  }
+
+  RAJA_HOST_DEVICE bool allocated() const { return mem != nullptr; }
+
+  RAJA_HOST_DEVICE value_type get(size_t i) const
+  {
+    return value_type(accessor::get(mem, i), accessor::get(mem_idx, i));
+  }
+  RAJA_HOST_DEVICE void set(size_t i, value_type val)
+  {
+    accessor::set(mem, i, first_type(val));
```

Comment:
@MrBurmark This `first_type(val)` call uses the `T()` operator in `ValLoc`. If there's a nice way to get rid of this, let me know.

Reply:
Can you use `val.getVal()`?

Reply:
Of course . . . thanks!